### PR TITLE
Remove branch pruning

### PR DESCRIPTION
**Pull request by norhh**
For now, I am using the args class. After next PR i will construct a new args class and send it to laser, that's cleaner.

### CODE REVIEW

**Comment by b-mueller** [comment]
Hmm, even with this branch I'm still getting the same "JUMPI  hiccups" with this [example](https://gist.github.com/b-mueller/a0834c19eac7e50d5b5f4addb4eb46e8). Run as:

```
$MYTH -v5 a SymbolicArbitrage.sol --disable-dependency-pruning
```

Will soon stop at some JUMPI for ~10 seconds:

```
mythril.laser.ethereum.instructions [DEBUG]: Evaluating JUMPI at 1286
````

File: mythril/support/support_args.py [hunk 1]
```diff
@@ -0,0 +1,13 @@
+from mythril.support.support_utils import Singleton
+
+
+class Args(object, metaclass=Singleton):
```
Comment (JoranHonig) [inline mythril/support/support_args.py]: I wouldn't have this be a singleton, this makes running multiple analysis within one process more complex

File: mythril/laser/ethereum/svm.py [hunk 2]
```diff
@@ -203,6 +203,14 @@ def _execute_transactions(self, address):
...
 
         for i in range(self.transaction_count):
+            if len(self.open_states) > 0:
+                old_states_count = len(self.open_states)
+                self.open_states = [
+                    state for state in self.open_states if state.constraints.is_possible
+                ]
+                prune_count = len(self.open_states) - old_states_count
+                if prune_count:
+                    log.info("Pruned {} states".format(prune_count))
```
Comment (JoranHonig) [inline mythril/laser/ethereum/svm.py]:
```suggestion
                    log.info("Pruned {} unreachable states".format(prune_count))
```

File: mythril/support/support_args.py [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+    """
+    This module helps in preventing args being sent through multiple of classes to reach any analysis/z
+    """
+
+    def __init__(self):
+        self.solver_timeout = 10000
+
+
+args = Args()
```
Comment (JoranHonig) [inline mythril/support/support_args.py]: if it's a singleton metaclass you don't have to pre-initialise it like this

File: mythril/support/model.py [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+import logging
+
+log = logging.getLogger(__name__)
+
+# LRU cache works great when used in powers of 2
+
+
+@lru_cache(maxsize=2 ** 23)
+def get_model(constraints, minimize=(), maximize=(), enforce_execution_time=True):
+    """
```
Comment (JoranHonig) [inline mythril/support/model.py]: perfect opportunity to add docs? :smile:

File: mythril/laser/ethereum/svm.py [hunk 5]
```diff
@@ -203,6 +203,14 @@ def _execute_transactions(self, address):
         self.time = datetime.now()
 
         for i in range(self.transaction_count):
+            if len(self.open_states) > 0:
```
Comment (JoranHonig) [inline mythril/laser/ethereum/svm.py]: I think this should be here regardless of branch pruning during execution

**Comment by b-mueller** [comment]
^^ now it's blazing fast :)

File: mythril/interfaces/cli.py [hunk 6]
```diff
@@ -446,6 +446,11 @@ def create_analyzer_parser(analyzer_parser: ArgumentParser):
         action="store_true",
         help="Don't attempt to retrieve contract code, variables and balances from the blockchain",
     )
+    options.add_argument(
+        "--deep-pruning",
```
[the remark this answers is not in the document]
Comment (norhh) [reply mythril/interfaces/cli.py]: Switched to `--sparse-pruning`

[the remark this answers is not in the document]
Comment (b-mueller) [reply mythril/interfaces/cli.py]: Wow, this discussion was 13 days ago? 🤣

**Pull request merged**